src/platform/redox/mod.rs

```diff
@@ -0,0 +1,554 @@
+                            });
+                        }
+                    },
+                    EventOption::Mouse(event) => callback(::Event::WindowEvent {
+                        window_id: ::WindowId(WindowId),
+                        event: ::WindowEvent::CursorMoved {
+                            device_id: ::DeviceId(DeviceId),
+                            position: (event.x, event.y).into(),
+                            modifiers: state.modifiers(),
+                        }
+                    }),
+                    EventOption::Button(event) => {
+                        while let Some((btn, btn_state)) = state.mouse(event.left, event.middle, event.right) {
+                            callback(::Event::WindowEvent {
+                                window_id: ::WindowId(WindowId),
+                                event: ::WindowEvent::MouseInput {
+                                    device_id: ::DeviceId(DeviceId),
+                                    state: btn_state,
+                                    button: btn,
+                                    modifiers: state.modifiers(),
+                                }
+                            });
+                        }
+                    },
+                    EventOption::Scroll(event) => callback(::Event::WindowEvent {
+                        window_id: ::WindowId(WindowId),
+                        event: ::WindowEvent::MouseWheel {
+                            device_id: ::DeviceId(DeviceId),
+                            delta: ::MouseScrollDelta::PixelDelta((event.x, event.y).into()),
+                            phase: ::TouchPhase::Moved,
+                            modifiers: state.modifiers(),
+                        }
+                    }),
+                    EventOption::Quit(_event) => callback(::Event::WindowEvent {
+                        window_id: ::WindowId(WindowId),
+                        event: ::WindowEvent::CloseRequested,
+                    }),
+                    EventOption::Focus(event) => callback(::Event::WindowEvent {
+                        window_id: ::WindowId(WindowId),
+                        event: ::WindowEvent::Focused(event.focused),
+                    }),
+                    EventOption::Move(event) => callback(::Event::WindowEvent {
+                        window_id: ::WindowId(WindowId),
+                        event: ::WindowEvent::Moved((event.x, event.y).into()),
+                    }),
+                    EventOption::Resize(event) => callback(::Event::WindowEvent {
+                        window_id: ::WindowId(WindowId),
+                        event: ::WindowEvent::Resized((event.width, event.height).into()),
+                    }),
+                    _ => ()
+                }
+            }
+        }
+    }
+
+    pub fn run_forever<F>(&mut self, mut callback: F)
+        where F: FnMut(::Event) -> ::ControlFlow,
+    {
+        // Yeah that's a very bad implementation.
```

Comment:
How will it be possible to improve it in the future?

Reply:
Not sure, I copied this specific function from the iOS module

Reply:
Oh ok.

I guess this is related to the use of `thread::sleep()`, likely because `self.poll_events()` is non-blocking on iOS. If this methods is a blocking call on Redox, the `sleep` call can likely be removed, and with that I don't see any obvious way why the implementation would be "bad".

Reply:
The methods are non-blocking because the window is opened with the Async WindowFlag

Reply:
Well in that case, the implementation could be improved by blocking (for example the wayland backend uses epoll to monitor the wayland socket) rather than loop-sleeping.